engine/env/spark/src/main/java/org/datacleaner/spark/utils/ResultFilePathUtils.java

```diff
@@ -19,86 +19,54 @@
-import java.net.URISyntaxException;
-import java.util.Date;
-
-import org.apache.hadoop.conf.Configuration;
-import org.apache.log4j.Logger;
+import org.apache.metamodel.util.Resource;
 import org.apache.spark.api.java.JavaSparkContext;
 import org.datacleaner.spark.SparkJobContext;
 import org.datacleaner.util.FileFilters;
 
-public class ResultFilePathUtils {
+import com.google.common.base.Strings;
 
-    private static final Logger logger = Logger.getLogger(ResultFilePathUtils.class);
+public class ResultFilePathUtils {
 
     private static final String DEFAULT_RESULT_PATH = "/datacleaner/results";
     private static final String RESULT_FILE_EXTENSION = FileFilters.ANALYSIS_RESULT_SER.getExtension();
 
     /**
-     * Gets the hdfs path of job's result. The path can be configured in the job
-     * properties file with property 'datacleaner.result.hdfs.path'. The path
-     * can be absolute(
+     * Gets the Resource to use for the job's result. The path can be configured
+     * in the job properties file with property 'datacleaner.result.hdfs.path'.
+     * The path can be absolute(
      * 'hdfs://[hostname]:[port]/myresults/myjob.analysis.result.dat') or
      * relative('/myresults/myjob.analysis.result.dat'). If no path is set in
      * the properties file and the system is configured to save the results, the
      * default path will be saved to
      * '/datacleaner/results/[jobname]-[timestamp].analysis.result.dat
      * 
-     * @throws URISyntaxException
-     * 
-     * @retur
+     * @return a Resource to use
      */
-    public static String getResultFilePath(final JavaSparkContext sparkContext, final SparkJobContext sparkJobContext)
-            throws URISyntaxException {
-        String resultPath = sparkJobContext.getResultPath();
-        final Configuration hadoopConfiguration = sparkContext.hadoopConfiguration();
-        /**
-         * The default value would be read from hadoop configuration at runtime
-         * from core-site.xml. It represents the machine's hostname and port.
-         * Example: hdfs://bigdatavm:9000
-         **/
-        final String fileSystemPrefix = hadoopConfiguration.get("fs.defaultFS");
-        if (resultPath == null || resultPath.isEmpty()) {
-            resultPath = createPath(fileSystemPrefix, DEFAULT_RESULT_PATH);
+    public static Resource getResultResource(final JavaSparkContext sparkContext,
+            final SparkJobContext sparkJobContext) {
+        final HdfsHelper hdfsHelper = new HdfsHelper(sparkContext);
+
+        final String resultPath;
+        if (Strings.isNullOrEmpty(sparkJobContext.getResultPath())) {
+            resultPath = DEFAULT_RESULT_PATH + '/' + generateResultFilename(sparkJobContext);
         } else {
-            final URI uri = URI.create(resultPath);
-            if (!uri.isAbsolute()) {
-                resultPath = createPath(fileSystemPrefix, resultPath);
+            if (hdfsHelper.isDirectory(sparkJobContext.getResultPath())) {
```

Comment:
Hmm after much thought, and while generally agreeing on your last statement (passing strings around is inferior to a proper type), here's my suggestion then: The problem needs to be solved not here, but then at the interface of SparkJobContext. This class returns the result path as a String, so the problem "starts" there. On the other hand, this remark was made in the Main class, which has a very string oriented interface (```main(String[] args)```) so in _this_ class I would never raise any alarm over handling strings because it's the intrinsic nature of a main class. But let's change it so that all the non-main classes (SparkJobContext in particular, and then also HdfsHelper I guess) deals with URI.
The other aspect to the discussion is about your general feeling about this in relation to loading of resources. And again I principally agree, but an important fact is missing: That the resources are loaded using a ```Converter``` implementation (ResourceConverter) and a Converter is _exactly_ for converting something to/from Strings (usually in relation to configured property values serialized in XML - but in this case also command line arguments). What we _could_ do is to add also a (non-interface) method to ResourceConverter which converts a URI to a Resource. That's totally OK with me, but it doesn't mean that the String-variant is not relevant - it really is, but just not for standard programmatic access like you see it, but for more polymorphic use (we have many Converters for all the different types of property values).